### ISSUE TITLE
git master fails to build 2019-01-05

### ISSUE THREAD

**Issue by uranther** [FStarLang/kremlin#124]
Hi - I am trying to use KreMLin for a new project and I am running into all sorts of basic issues to get it running. Starting with a fresh `git clone`, it fails to `make` and `make test`.

```
$ echo $FSTAR_HOME
/home/james/Code/FStar
$ echo $KRML_HOME
/home/james/Code/kremlin
$ cd $FSTAR_HOME && git rev-parse HEAD
9215037e86a9664f8983d95467da7366662b81ec
```

Here's the last output of `make`:

```
[clipped]
...
⚙ KreMLin auto-detecting tools. Here's what we found:
readlink is: readlink
KreMLin called via: ../krml
the Kremlin executable is: /home/james/Code/kremlin/_build/src/Kremlin.native
KreMLin home is: /home/james/Code/kremlin
⚙ KreMLin will drive F*. Here's what we found:
read FSTAR_HOME via the environment
fstar is on branch master
fstar is: /home/james/Code/FStar/bin/fstar.exe /home/james/Code/kremlin/runtime/WasmSupport.fst /home/james/Code/FStar/ulib/FStar.UInt128.fst --trace_error
--cache_checked_modules --expose_interfaces --include /home/james/Code/kremlin/kremlib --include /home/james/Code/kremlin/include                          
✔ [PrettyPrinting] ⏱️ 5ms
KreMLin: wrote out .c files for FStar_UInt_8_16_32_64, FStar_ModifiesGen
KreMLin: wrote out .h files for Prims, FStar_BitVector, FStar_UInt, FStar_UInt_8_16_32_64, FStar_UInt128, C_Endianness, FStar_Int_Cast_Full, TestLib, C_Failure, FStar_HyperStack_IO, FStar_Bytes, C_String, FStar_String, FStar_Char, C, FStar_IO, FStar_Float, FStar_Int_Cast, FStar_Int, FStar_Int8, FStar_Int32, FStar_Int64, FStar_Int63, FStar_All, FStar_Monotonic_Heap, FStar_ST, FStar_UInt63, FStar_Kremlin_Endianness, C_Loops, FStar_Monotonic_HyperStack, LowStar_Buffer, LowStar_Monotonic_Buffer, FStar_ModifiesGen, FStar_Date, FStar_HyperStack_All, FStar_HyperStack_ST, FStar_Monotonic_HyperHeap, FStar_Int16, FStar_Math_Lib, FStar_Mul, FStar_Pervasives
cp c/*.c dist/generic/
make -C dist/generic -f Makefile.basic
make[2]: Entering directory '/home/james/Code/kremlin/kremlib/dist/generic'
In file included from fstar_int8.c:4:
FStar_Int8.h:13:10: fatal error: kremlib.h: No such file or directory
 #include "kremlib.h"
          ^~~~~~~~~~~
compilation terminated.
Makefile.basic:44: fstar_int8.d: No such file or directory
make[2]: *** [Makefile.basic:39: fstar_int8.d] Error 1
make[2]: Leaving directory '/home/james/Code/kremlin/kremlib/dist/generic'
make[1]: *** [Makefile:121: compile-dist-generic] Error 2
make[1]: Leaving directory '/home/james/Code/kremlin/kremlib'
make: *** [Makefile:28: kremlib] Error 2
```

If I `cp -R include/*` into each of the `dist/` directories (`minimal`, `generic`, `uint128`), I can get it to build `libkremlib.a` successfully. Now `make test` fails with missing JSON data:

```
[clipped]
...
Verified module: FStar.Int16 (1520 milliseconds)
All verification conditions discharged successfully
fstar.exe --cache_checked_modules --use_two_phase_tc true --cache_dir .cache --odir .output --include hello-system --include ../kremlib/compat --include ../kremlib --include ../runtime --include /home/james/Code/FStar/examples/low-level/crypto --include ../book --use_hints --record_hints --hint_info --hint_file .hints/FStar.Int8.fst.hints /home/james/.opam/system/lib/fstar/FStar.Int8.fst && \
touch .cache/FStar.Int8.fst.checked

Unexpected error; please file a bug report, ideally with a minimized version of the source program that triggered the error.
Yojson.Json_error("Blank input data")
make[1]: *** [Makefile:75: .cache/FStar.Int8.fst.checked] Error 1
make[1]: Leaving directory '/home/james/Code/kremlin/test'
make: *** [Makefile:50: test] Error 2
```

**Comment by msprotz** [comment]
Hello,

Thanks for the bug report. A recent round of changes took a dependency on an environment variable, KREMLIN_HOME, that as you may guess points to the root of the kremlin directory. I've updated the one Makefile that depends on this to properly error out with a helpful message.

Regarding the Int8 error, this is caused by a blank file, namely test/.hints/FStar.Int8.fst.hints; however, I didn't have this error on my system, even with that blank file. Can you give me a little more info? Namely: OCaml version, Yojson package version, F* version.

Thanks!

**Comment by uranther** [comment]
Appreciate the swift response! Setting `$KREMLIN_HOME` fixes the build, but yes, still getting the same failure on `make test`.

* OCaml version 4.05.0
* YoJSON 1.3.3
* FStar 0.9.7.0~dev (9215037e86a9664f8983d95467da7366662b81ec)

**Comment by uranther** [comment]
It seems that 1.5.0 is not listed as an available version in opam. I upgraded it to 1.4.1 and after a `make distclean` and `make test` it fails in the same way.

**Comment by msprotz** [comment]
Yojson 1.5.0 is available there: https://opam.ocaml.org/packages/yojson/ -- try opam update first?

Then, can you try removing the empty .hints file?

Thanks!

Jonathan

**Comment by uranther** [comment]
Alright - Yojson 1.5.0 is only available on opam 2.0 and that is not yet available in Gentoo Portage. After adding the `ml-overlay`, rebuilding OCaml, opam, FStar and KreMLin fresh, (and making a softlink to `fstar.exe` since the Makefile doesn't accept `fstar`) I get a completely different error on `make test`:

```
[clipped]
...
./Tests.native
Printing with w=80
typedef uint8_t *t[4U]
typedef uint8_t (*t)[4U]
typedef uint8_t (*(**t)(uint32_t x0))[3U]
typedef struct  { uint8_t bar; } (**t)(uint32_t x0)
uint8_t (*f(uint8_t (*x0)(uint8_t x0)))(uint8_t x0)
{
  KRML_HOST_PRINTF("KreMLin abort at %s:%d\n%s\n",
    __FILE__,
    __LINE__,
    "test");
  KRML_HOST_EXIT(255U);
}
(C11.Function ([], false,
   ([], (C11.Int Constant.UInt8), None,
    [((C11.Function (None,
         (C11.Pointer ([],
            (C11.Function (None, (C11.Ident "f"),
               [([], (C11.Int Constant.UInt8),
                 (C11.Function (None, (C11.Pointer ([], (C11.Ident "x0"))),
                    [([], (C11.Int Constant.UInt8), (C11.Ident "x0"))])))
                 ]
               ))
            )),
         [([], (C11.Int Constant.UInt8), (C11.Ident "x0"))])),
      None)]),
   (C11.Compound
      [(C11.Expr
          (C11.Call ((C11.Name "KRML_HOST_PRINTF"),
             [(C11.Literal "KreMLin abort at %s:%d\\n%s\\n");
               (C11.Name "__FILE__"); (C11.Name "__LINE__");
               (C11.Literal "test")]
             )));
        (C11.Expr
           (C11.Call ((C11.Name "KRML_HOST_EXIT"),
              [(C11.Constant (Constant.UInt8, "255"))])))
        ])
   ))
env = hello, TUnit
env = hello, EBound: bool = 0
make -C test
make[1]: Entering directory '/home/james/Code/kremlin/test'
fstar.exe --cache_checked_modules --use_two_phase_tc true --cache_dir .cache --odir .output --include hello-system --include ../kremlib/compat --include ../kremlib --include ../runtime --include /home/james/Code/FStar/examples/low-level/crypto --include ../book --use_hints --record_hints --hint_info --already_cached 'FStar C TestLib Spec.Loops -C.Compat -C.Nullity' --codegen Kremlin \
  --extract_module  \
  
last option 'extract_module' takes an argument but has none
make[1]: *** [Makefile:80: .output/FStar_UInt128.krml] Error 1
make[1]: Leaving directory '/home/james/Code/kremlin/test'
make: *** [Makefile:50: test] Error 2
```

**Comment by msprotz** [comment]
do you have the most recent versions of both F* and kremlin?

**Comment by uranther** [comment]
Yes. Here's my version output:

```
$ which -a fstar                                                                                                                                   
fstar: aliased to fstar --n_cores 8                                                                                                                        
/home/james/.local/bin/fstar
$ ls -la ~/.local/bin/fstar                       
lrwxrwxrwx 1 james james 36 Oct  5 18:24 /home/james/.local/bin/fstar -> /home/james/Code/FStar/bin/fstar.exe
$ fstar --version                                                                                                                                  
F* 0.9.7.0~dev                                                                                                                                             
platform=Linux_x86_64      
compiler=OCaml 4.07.1     
date=2019-01-10T13:56:31-0500                                                                                                                              
commit=43bc5735e0f0590f44a39d5f8189db772957bc94 (dirty)

$ which krml
/home/james/.local/bin/krml
$ ls -la ~/.local/bin/krml
lrwxrwxrwx 1 james james 29 Oct  5 18:01 /home/james/.local/bin/krml -> /home/james/Code/kremlin/krml
```

and, well, `krml` has no `-version` option so here's the commit hash

```
$ cd $KREMLIN_HOME && git rev-parse HEAD
e9a42a8091a53f8cd4c17ffb38e259125ecc56c8
```

**Comment by msprotz** [comment]
could it be that you have a stale .depend in your test directory?

**Comment by beurdouche** [comment]
Just to be safe, can you run everything without the `--n_cores` option : )

**Comment by uranther** [comment]
`make test` seems to have completely successfully!

Well, I had to reboot my machine for another reason, so that means I started with a fresh shell environment. I went ahead and removed the `--n-cores` alias even though `fstar.exe` was being executed directly by the make script. `.depend` had nothing in it but I deleted it anyway.

Messing with the .hints files didn't seem to fix anything, and this time around I started from a fresh KreMLin repo. Although now after everything is said and done, git shows modifications to the .hints both in `kremlib/` and `test/`
